### PR TITLE
MWPW-155940: MAS AEM client 

### PR DESCRIPTION
**Pull request by yesil**
aem.js in mas is a partial client for several of the Sites APIs (https://developer.adobe.com/experience-cloud/experience-manager-apis/api/stable/sites/) that will support merch-card coming from Odin, as well as authoring of merch-cards in MAS Studio.

Also, this PR adds a github workflow to integrate MAS feature(libs/features/mas) unit tests to be run on PRs and send codecov reports.

Resolves: 
- [MWPW-155941](https://jira.corp.adobe.com/browse/MWPW-155941)
- [MWPW-156983](https://jira.corp.adobe.com/browse/MWPW-156983)
- [MWPW-157404](https://jira.corp.adobe.com/browse/MWPW-157404)
- [MWPW-157405](https://jira.corp.adobe.com/browse/MWPW-157405)

**Test URLs:**
- Before: https://main--milo--adobecom.hlx.page/drafts/ilyas/merch-card?martech=off
- After: https://mwpw-155940--milo--yesil.hlx.page/drafts/ilyas/merch-card?martech=off

Note: this PR replaces https://github.com/adobecom/milo/pull/2761 (closed)

### CODE REVIEW

File: libs/features/mas/web-components/src/global.css.js
Comment (3ch023) [inline libs/features/mas/web-components/src/global.css.js:612]: this might affect someone in Milo who used sp-button outside of Merch team
Maybe we can change it to `sp-button a[is="checkout-link"]`?

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (3ch023) [inline libs/features/mas/web-components/src/merch-datasource.js:52]: We should discuss if it makes sense to split merch-datasource.
It is the same situation as with merch-card block in milo it will keep growing with each card variant added and might become too complicated.

**Review by 3ch023 — APPROVED** [review]
approving, 2 minor comments

File: libs/features/mas/mas/src/mas.js
Comment (npeltier) [inline libs/features/mas/mas/src/mas.js:14]: but this logic (and URL) is already in merch.js, no? i guess we should factorize this

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (npeltier) [reply libs/features/mas/web-components/src/merch-datasource.js:52]: and typically mapping could be held, as well as css in one given variant file

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (npeltier) [inline libs/features/mas/web-components/src/merch-datasource.js:256]: what does that do?

Comment (npeltier) [reply libs/features/mas/web-components/src/merch-datasource.js:256]: i understand race & resolve but i don't understand the logic here

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (npeltier) [inline libs/features/mas/web-components/src/merch-datasource.js:248]: what is the reasoning behind those refs + removal? to remove readd everything but datasource child?

**Review by npeltier — APPROVED** [review]

File: libs/features/mas/web-components/src/aem.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/aem.js:24]: Should we include error handling in some or every `fetch` to manage network errors or unexpected responses?

File: libs/features/mas/web-components/src/aem.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/aem.js:125]: Should we consider caching the CSRF token to improve performance? Wondering if this changes frequently?

Comment (Axelcureno) [reply libs/features/mas/web-components/src/aem.js:125]: Or I guess if we won't copy fragments too many times it might be okay...

File: libs/features/mas/web-components/src/merch-card.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/merch-card.js:534]: Curious of why only in special offers and ccdAction is this added?

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/merch-datasource.js:85]: I think we can remove L85, L86.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/merch-datasource.js:74]: What if we use constants for the variant names instead of magic strings to avoid typos and make the code easier to maintain? i.e.
```suggestion
const VARIANTS = {
    CATALOG: 'catalog',
    AH: 'ah',
    CCD_ACTION: 'ccd-action',
    SPECIAL_OFFERS: 'special-offers',
};

const cardContent = {
    [VARIANTS.CATALOG]: {
        title: { tag: 'h3', slot: 'heading-xs' },
        prices: { tag: 'h3', slot: 'heading-xs' },
        description: { tag: 'div', slot: 'body-xs' },
        ctas: { size: 'l' },
    },
    [VARIANTS.AH]: {
        title: { tag: 'h3', slot: 'heading-xxs' },
        prices: { tag: 'h3', slot: 'heading-xs' },
        description: { tag: 'div', slot: 'body-xxs' },
        ctas: { size: 's' },
    },
    [VARIANTS.CCD_ACTION]: {
        title: { tag: 'h3', slot: 'heading-xs' },
        prices: { tag: 'h3', slot: 'heading-xs' },
        description: { tag: 'div', slot: 'body-xs' },
        ctas: { size: 'l' },
    },
    [VARIANTS.SPECIAL_OFFERS]: {
        name: { tag: 'h4', slot: 'detail-m' },
        title: { tag: 'h4', slot: 'detail-m' },
        backgroundImage: { tag: 'div', slot: 'bg-image' },
        prices: { tag: 'h3', slot: 'heading-xs' },
        description: { tag: 'div', slot: 'body-xs' },
        ctas: { size: 'l' },
    },
};
```

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/merch-datasource.js:143]: Seems like this can be `const`

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (Axelcureno) [reply libs/features/mas/web-components/src/merch-datasource.js:52]: I like this idea, files are getting too bulky and harder to maintain.

File: libs/features/mas/web-components/src/utils.js
Comment (Axelcureno) [inline libs/features/mas/web-components/src/utils.js:21]: This function looks similar to Milo's. Are we recreating it to decouple the 2 codebases?

**Review by Axelcureno — APPROVED** [review]

**Comment by Roycethan** [comment]
@yesil I see some of the catalog issues fixed earlier are back in this brnach...
ex: https://main--cc--adobecom.hlx.live/products/catalog?milolibs=mwpw-155940--milo--yesil
1) Click on Filter "Mobile" and uncheck - results should be back to normal, the state is not reverted to original
2) Click on Sorting say Alphabetical and again click back to Popularity - the sorting is not changed back
3) when url is launched the "Show more" button is not rendered
These are all fixed in https://github.com/adobecom/milo/pull/2653 back by @mirafedas , Can you plz check if the latest is in or anything missing here ?
cc @Axelcureno  @3ch023  @npeltier @afmicka

File: libs/features/mas/web-components/src/global.css.js
Comment (yesil) [reply libs/features/mas/web-components/src/global.css.js:612]: although no one should have sp-button a in theory and this is not elegant, but the only way to have checkout links that look like a CTA without consonant styles.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-datasource.js:52]: this will be addressed post PR merge as tech-debt.

File: libs/features/mas/web-components/src/aem.js
Comment (yesil) [reply libs/features/mas/web-components/src/aem.js:3]: As discussed in huddle,  we are using only the open AEM APIs and Odin happens to be the AEM instance behind the scene.

File: libs/features/mas/mas/src/mas.js
Comment (yesil) [reply libs/features/mas/mas/src/mas.js:14]: this is needed for mas.js.  We can discuss about merch.js when we consider consuming mas.js in Milo.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-datasource.js:290]: @npeltier later in the file we have  `get updateComplete()` which is a promise that resolves when datasource job is done.
`readyPromise`  once pending, should not be re-initialized, otherwise some caller might end up awaiting for ever.
here, if we are fetching the data already, we skip refreshing.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-datasource.js:282]: @npeltier MAS studio refreshes datasource un updates.  merch-card can have elements in the default slot that are not created by the datasource but by the consumer. e.g: mas studio adds a `sp-status-light` to display the publish status. Basically, when we refresh a datasource, we only clear and re-create the elements that are tracked.

File: libs/features/mas/web-components/src/aem.js
Comment (yesil) [reply libs/features/mas/web-components/src/aem.js:24]: Ideally yes. This is needed in MAS Studio only and actually a temporary solution, and needs to be removed once the AEM eng fixes a bug I found in the CF cloning.
Also, error handling is implemented on the consumer of aem.js since the only choice we have in aem.js is to throw an Error.

File: libs/features/mas/web-components/src/aem.js
Comment (yesil) [reply libs/features/mas/web-components/src/aem.js:125]: csrf token changes often and as said before, we should not use `copyFragmentClassic` but `copyFragment` once the API is fixed.

File: libs/features/mas/web-components/src/merch-card.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-card.js:534]: Ideally we need it in all the cards but I don't want to add it to avoid breaking them for no reason.
currently those two are used in MAS Studio, where default slot is used.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-datasource.js:177]: removed the line and adjusted the code.

File: libs/features/mas/web-components/src/utils.js
Comment (yesil) [reply libs/features/mas/web-components/src/utils.js:21]: I don't want to pull Milo's utils.js in mas.js that we ship to non Milo consumers as well.

**Comment by yesil** [comment]
@Roycethan thanks for reporting the regression, reverted/fixed the change, thx.

File: libs/features/mas/web-components/src/global.css.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/global.css.js:1089]: I don't think this is needed since you already declared on L#613

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (narcis-radu) [inline libs/features/mas/web-components/src/merch-datasource.js:227]: I see multiple occurrences, but I'll only comment here: the preferable way to write this is `if (something) doSomething();`

**Review by narcis-radu — CHANGES_REQUESTED** [review]
I have some doubts about content being uploaded in the repository and folder structure so I will request some changes for now.

File: .github/workflows/run-mas-tests.yaml
Comment (yesil) [reply .github/workflows/run-mas-tests.yaml:5]: I wondered about the same but I simply copied the one of Milo.
.github/workflows/run-tests.yaml

File: libs/deps/mas/commerce.js
Comment (yesil) [reply libs/deps/mas/commerce.js:4]: I don't remember it, do you have any pointer to the discussion?

File: libs/deps/mas/mas.js
Comment (yesil) [reply libs/deps/mas/mas.js:166]: Please do not review deps/mas/*

this src of this has already been reviewed in previous PRs. It is css scoped inside shadow-root, so no impact outside.

File: libs/deps/mas/merch-card-collection.js
Comment (yesil) [reply libs/deps/mas/merch-card-collection.js:151]: I'll remove sourcemap from this but which Safari error? I'm not aware of it.

File: libs/features/mas/mas/src/mas.js
Comment (yesil) [reply libs/features/mas/mas/src/mas.js:8]: mas.js will be used by CCD/CCH first and we are about to enable this feature.

File: libs/features/mas/mas/src/mas.js
Comment (yesil) [reply libs/features/mas/mas/src/mas.js:15]: This should be considered as an external resource similar to prices.
cc: @npeltier wdyt?

File: libs/features/mas/mocks/aem.js
Comment (yesil) [reply libs/features/mas/mocks/aem.js:2]: Good point, I don't remember how they were in the adobecom/mas repo. We are still shuffling things around in the milo repository after moving mas to Milo features. I'll update.

File: libs/features/mas/web-components/assets/special-offers-cc.png
Comment (yesil) [reply libs/features/mas/web-components/assets/special-offers-cc.png:1]: we have test content that references images. Since we don't want external calls, this seems fair to me.
What is the reason for not having test images in the repository?

File: libs/features/mas/web-components/src/global.css.js
Comment (yesil) [reply libs/features/mas/web-components/src/global.css.js:1089]: this file has grown too fast too big. We are going to split it.
merged `sp-button a` rules under the same rule.

File: libs/features/mas/mocks/aem.js
Comment (yesil) [reply libs/features/mas/mocks/aem.js:2]: I think we need to discuss if we want to reflect the whole MAS hierarchy to milo/test/features/mas.
MAS is a multimodule project for now. That would mean too much change for no justified gain.
My preference is to keep MAS and related tests file where they are.
We will discuss this in our dev meeting but I'll keep it out of the scope of this PR.

File: libs/features/mas/web-components/src/global.css.js
Comment (yesil) [reply libs/features/mas/web-components/src/global.css.js:1077]: @3ch023 I change the rule as above, since merch-cards can have regular CTAs with links as well.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (narcis-radu) [reply libs/features/mas/web-components/src/merch-datasource.js:232]: We need to agree on a consistent format. If (when) somebody else picks up the work, they should write the code in a similar matter.

File: libs/features/mas/web-components/assets/special-offers-cc.png
Comment (narcis-radu) [reply libs/features/mas/web-components/assets/special-offers-cc.png:1]: In that case, assets should be in a test folder, see a related comment https://github.com/adobecom/milo/pull/2798#discussion_r1736269142

File: libs/features/mas/mas/src/mas.js
Comment (narcis-radu) [reply libs/features/mas/mas/src/mas.js:8]: let's make it part of a new PR then

File: libs/deps/mas/merch-card-collection.js
Comment (narcis-radu) [reply libs/deps/mas/merch-card-collection.js]: I'll share the link internally

File: libs/deps/mas/mas.js
Comment (narcis-radu) [reply libs/deps/mas/mas.js:166]: it was a personal curiosity, thanks for confirming

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (yesil) [reply libs/features/mas/web-components/src/merch-datasource.js:232]: This is MAS engineering preference. w.r.t `if`s,  Milo codebase is not consistent already. If you want to enforce a coding style, please start with a Github discussion. Then each them can address it at their pace.

File: libs/features/mas/web-components/assets/special-offers-cc.png
Comment (yesil) [reply libs/features/mas/web-components/assets/special-offers-cc.png:1]: I agree we should move test images to test folder.

`assets` folder was created by the visual tests of TWP(on hold, not finished) which uses projectRoot/assets as base for the baseline images. 

Removed this image since we already have it in test/img

We will revisit `assets` when the TWP work will be resumed.

**Comment by yesil** [comment]
@narcis-radu I think I addressed all your feedbacks. Let me know if you still have questions.

**Comment by yesil** [comment]
@Roycethan can you verify once more? thanks.

File: libs/features/mas/mas/src/mas.js
Comment (npeltier) [reply libs/features/mas/mas/src/mas.js:14]: yes indeed. Literals new data testing could be considered but i would rather do this in another (priorized) ticket.

**Comment by npeltier** [comment]
hey @narcis-radu can you summarise what are belong your comments the requested changes?

**Comment by yesil** [comment]
@narcis-radu as discussed, tracked follow-up items in https://jira.corp.adobe.com/browse/MWPW-157565.

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (npeltier) [reply libs/features/mas/web-components/src/merch-datasource.js:230]: @narcis-radu am i right to understand this is not a requested change? i understand the need for coding rules, but i guess we should probably discuss that in a broader place before making them requested (discussion?)
also i like additional info in the comment to differentiate the two types of comments (minor or requested change), when outcome status is "requested changes"

**Comment by Roycethan** [comment]
Done smoke tests post code review changes. and adding Ready for stage since Narcis's comments are tracked in separate jira for follow-up
cc @yesil

**Review by narcis-radu — APPROVED** [review]
issues will be tracked in a separate ticket

File: .github/workflows/run-mas-tests.yaml
Comment (mokimo) [reply .github/workflows/run-mas-tests.yaml:5]: Hmm... it's running once we push to main, we don'nt need that at all.
Also, not for the existing workflow. Since we already run the tests on each PR, no need to run it after something is merged

File: libs/features/mas/web-components/src/merch-datasource.js
Comment (narcis-radu) [reply libs/features/mas/web-components/src/merch-datasource.js:237]: @narcis-radu - just a nit (for now).
@yesil - All code should follow the Milo guidelines. The patterns have been silently enforced so far, but there are some misses since PRs have been approved when this was the only nit (something similar to this PR, right?).
I have created a thread to discuss preferences - https://github.com/orgs/adobecom/discussions/2813

File: libs/features/mas/mas/src/mas.js
Comment (narcis-radu) [reply libs/features/mas/mas/src/mas.js:14]: @npeltier - do you have a follow-up JIRA ticket?

Comment (npeltier) [reply libs/features/mas/mas/src/mas.js:14]: https://jira.corp.adobe.com/browse/MWPW-157607

File: libs/features/mas/web-components/src/aem.js
Comment (robert-bogos) [inline libs/features/mas/web-components/src/aem.js:88]: Should we add a check for the response in here?
```suggestion
    async getFragment(res) {
    if(res.ok) {
          const eTag = res.headers.get('Etag');
          const fragment = await res.json();
          fragment.etag = eTag;
          return fragment;
      } 
       throw new Error('Failed to get fragment by id');
    }
```

File: libs/features/mas/web-components/src/aem.js
Comment (robert-bogos) [inline libs/features/mas/web-components/src/aem.js:16]:
```suggestion
            // IMS users might not have all the permissions, token in the sessionStorage is a temporary workaround
```

**Comment by yesil** [comment]
@robert-bogos fyi, I covered all fetch w.r.t response status and network errors.

**Comment by yesil** [comment]
I removed Ready-For-Stage label to avoid affecting the upcoming catalog release.
Will re-add it back with @afmicka post go-live.

**Pull request merged**
